dask/array/tests/test_array_function.py

```diff
@@ -0,0 +1,100 @@
+import pytest
+np = pytest.importorskip('numpy', minversion='1.16')
+
+import dask.array as da
+from dask.array.utils import assert_eq
+
+
+functions = [
+    lambda x: np.concatenate([x, x, x]),
+    lambda x: np.cov(x, x),
+    lambda x: np.dot(x, x),
+    lambda x: np.dstack(x),
+    lambda x: np.flip(x, axis=0),
+    lambda x: np.hstack(x),
+    lambda x: np.matmul(x, x),
+    lambda x: np.mean(x),
+    lambda x: np.stack([x, x]),
+    lambda x: np.sum(x),
+    lambda x: np.var(x),
+    lambda x: np.vstack(x),
+    lambda x: np.fft.fft(x.rechunk(x.shape) if isinstance(x, da.Array) else x),
+    lambda x: np.fft.fft2(x.rechunk(x.shape) if isinstance(x, da.Array) else x),
+    lambda x: np.linalg.norm(x)
+]
+
+
+notimpl_functions = [
+    lambda x: np.min_scalar_type(x),
+    lambda x: np.linalg.det(x),
+    lambda x: np.linalg.eigvals(x)
+]
+
+
+@pytest.mark.parametrize('func', functions)
+def test_array_function_dask(func):
+    x = np.random.random((100, 100))
+    y = da.from_array(x, chunks=(50, 50))
+    res_x = func(x)
+    res_y = func(y)
+
+    assert isinstance(res_y, da.Array)
+    assert_eq(res_y, res_x)
+
+
+@pytest.mark.parametrize('func', unimpl_functions)
+def test_array_notimpl_function_dask(func):
+    x = np.random.random((100, 100))
+    y = da.from_array(x, chunks=(50, 50))
+
+    with pytest.raises(TypeError) as e:
+        res_y = func(y)
+
+
+def test_array_function_sparse_transpose():
+    sparse = pytest.importorskip('sparse')
+    x = da.random.random((500, 500), chunks=(100, 100))
+    x[x < 0.9] = 0
+
+    y = x.map_blocks(sparse.COO)
+
+    xT = da.transpose(x).compute()
+    yT = da.transpose(y).compute()
+
+    assert_eq(np.transpose(x), xT)
+    assert_eq(np.transpose(y), yT.todense())
+
+
+@pytest.mark.xfail(reason="requires sparse support for __array_function__",
+                   strict=False)
+def test_array_function_sparse_tensordot():
+    sparse = pytest.importorskip('sparse')
+    x = np.random.random((2, 3, 4))
+    x[x < 0.9] = 0
+    y = np.random.random((4, 3, 2))
+    y[y < 0.9] = 0
+
+    xx = sparse.COO(x)
+    yy = sparse.COO(y)
+
+    assert_eq(np.tensordot(x, y, axes=(2, 0)),
+              np.tensordot(xx, yy, axes=(2, 0)).todense())
+
+
+def test_array_function_cupy_svd():
+    cupy = pytest.importorskip('cupy')
+    x = cupy.random.random((500, 100))
+
+    y = da.from_array(x, chunks=(100, 100), asarray=False)
+
+    u_base, s_base, v_base = da.linalg.svd(y)
+    u, s, v = np.linalg.svd(y)
+
+    # Using single-threaded for now, multi-threaded depends on
+    # https://github.com/cupy/cupy/pull/2053
+    assert_eq(u.compute(scheduler='single-threaded'),
+              u_base.compute(scheduler='single-threaded'))
+    assert_eq(s.compute(scheduler='single-threaded'),
+              s_base.compute(scheduler='single-threaded'))
+    assert_eq(v.compute(scheduler='single-threaded'),
+              v_base.compute(scheduler='single-threaded'))
```

Comment:
Note that `assert_eq` runs things single threaded anyway.  We also prefer that people not pre-compute results, that way `assert_eq` can run some additional checks on the dask object.

Reply:
> Note that assert_eq runs things single threaded anyway.

What's the reasoning behind that? I particularly think this is bad, since it wouldn't test potential issues that arise from parallel execution (e.g., race conditions), or do you have tests somewhere that actually do the multi-threaded testing?

Reply:
We do have a variety of tests for each of the schedulers, yes (the dask/distributed test suite in particular is quite extensive).  However testing each of the schedulers against each of the collections (arrays, bags, delayed, dataframes, futures, ml, ...) probably isn't necessary, and tends to make debugging failures difficult. 

The separation between collections and schedulers via task graphs ends up being clean enough that single-threaded testing for the collections seems to have won out over time as the most pragmatic choice.

Regardless, even if we were to make a change, this PR probably isn't the place to make it.  We should continue to insert collections directly into assert_eq, and then if we want to change the default scheduler for all tests we can do it there in the future.

Reply:
Thanks for the details, @mrocklin.